internal/pkg/agent/application/dispatcher/dispatcher.go

```diff
@@ -86,6 +95,17 @@ func (ad *ActionDispatcher) Dispatch(ctx context.Context, acker acker.Acker, act
 		span.End()
 	}()
 
+	actions = ad.queueScheduledActions(actions)
+	actions = ad.dispatchCancelActions(ctx, actions, acker)
+	queued, expired := ad.gatherQueuedActions(time.Now().UTC())
+	ad.log.Debugf("Gathered %d actions from queue, %d actions expired", len(queued), len(expired))
+	ad.log.Debugf("Expired actions: %v", expired)
+	actions = append(actions, queued...)
+
+	if err := ad.queue.Save(); err != nil {
+		return fmt.Errorf("failed to persist action_queue: %w", err)
```

Comment:
What happens here? If it fails to save will the Elastic Agent try to get those actions again from Fleet Server or our they now lost because it failed to save?

Understanding the flow here is really important.

Reply:
👍 i'll look into it tomorrow

Reply:
Before the dispatcher run, the gateway will gather and and attempt to persist the `ackToken` - if this fails it logs a message; at this point the `ackToken` in memory will be more recent than the one on disk. On the next checkin only new actions are retrieved, however if the agent needs to restart (after saving to disk fails) then the checkin will get all actions since the saved `ackToken` and may retry actions.

I've changed queue saving to log an error message so it follows the same behaviour. If persisting the queue fails then the in-memory queue will be more recent than the on-disk queue; if the agent restarts then all scheduled actions will be re-added to the queue and schedueled+reran/cancelled/expired as expected.
However there may be an edge case if the `ackToken` is saved and the queue fails, in this case scheduled actions may be lost if the agent is reset.